packages/commonwealth/client/scripts/views/components/ReactNativeBridge/ReactNativeScrollToTopListener.tsx

```diff
@@ -0,0 +1,65 @@
+import type { ReactNativeWebView } from 'hooks/useReactNativeWebView';
+import { useCallback, useEffect } from 'react';
+
+declare global {
+  interface Window {
+    ReactNativeWebView?: ReactNativeWebView;
+  }
+}
+
+type ScrollToTop = {
+  type: 'scroll-to-top';
+  link: string;
+};
+
+function isScrollToTop(data: object): data is ScrollToTop {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return (data as any).type === 'scroll-to-top';
+}
+
+/**
+ * Handles routing request on the frontend to new URLs from requests from
+ * react-native
+ */
+export const ReactNativeScrollToTopListener = () => {
+  const handleMessage = useCallback((message: MessageEvent) => {
+    const obj = messageToObject(message.data);
+    if (obj && typeof message.data === 'object') {
+      if (isScrollToTop(obj)) {
+        smoothScrollAllToTop();
+      }
+    }
+  }, []);
+
+  useEffect(() => {
+    window.addEventListener('message', handleMessage);
+
+    return () => {
+      window.removeEventListener('message', handleMessage);
+    };
+  }, [handleMessage]);
+
+  return null;
+};
+
+function messageToObject(message: string | object): object | null {
+  try {
+    return typeof message === 'string' ? JSON.parse(message) : message;
+  } catch (e) {
+    // this could happen if another library is sending non-JSON data via
+    // postMessage
+    return null;
+  }
+}
+
+function smoothScrollAllToTop() {
+  // Scroll the window itself to the top
+  window.scrollTo({ top: 0, behavior: 'smooth' });
+
+  // Scroll all scrollable elements to the top smoothly
+  document.querySelectorAll('*').forEach((el) => {
```

Comment:
ui performance concerns? It selects every single element on the page. When "handleMessage" is being invoked or how often?

Reply:
Almost never... it only happens when the user clicks the top bar to scroll to the top.